{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Reading the document and abstraction of gene information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[{'count': 1, 'A': 20, 'C': 33, 'T': 25, 'G': 48, 'Total': 126, 'C-G-ratio': 64.28571428571429}, {'count': 2, 'A': 54, 'C': 33, 'T': 37, 'G': 53, 'Total': 177, 'C-G-ratio': 48.58757062146893}, {'count': 3, 'A': 323, 'C': 157, 'T': 345, 'G': 202, 'Total': 1027, 'C-G-ratio': 34.95618305744888}]\n",
      "[{'count': 1, 'A': 210, 'C': 282, 'T': 252, 'G': 340, 'Total': 1084, 'C-G-ratio': 57.38007380073801}, {'count': 2, 'A': 40, 'C': 20, 'T': 35, 'G': 29, 'Total': 124, 'C-G-ratio': 39.516129032258064}]\n"
     ]
    }
   ],
   "source": [
    "with open ('CYCS_dog.txt','r') as file_1:\n",
    "    string = file_1.read()\n",
    "    \n",
    "    #forming a list out of the file given sequence code\n",
    "    list_strg_1 = string.split('<', 6)\n",
    "    list_strg_1.pop(0)\n",
    "    \n",
    "    #creating tuples inside list with each part(eg. exon,intron) as a element of the list\n",
    "    exon_intron_tuple = []\n",
    "    for i in enumerate(list_strg_1):\n",
    "        exon_intron_tuple.append(i)\n",
    "   \n",
    "    #removal of the unwanted CDS gene segment\n",
    "    exon_intron_tuple.pop(0)\n",
    "    \n",
    "    #creating exon and intron list to store respective values seperately\n",
    "    exon_list = []\n",
    "    intron_list = []\n",
    "    \n",
    "    \n",
    "    #counting number of exons and introns in the sequence and storing them in respective list seperately\n",
    "    exon_num = 0\n",
    "    intron_num = 0\n",
    "    for item in exon_intron_tuple:\n",
    "        if 'exon' in item[1]:\n",
    "            exon_num += 1    \n",
    "            ex = {}\n",
    "\n",
    "            #modifying the string in the exon for removal of non-nucleotides\n",
    "            exon_region = [char for char in item[1] if char == 'A' or char == 'T' or char == 'G' or char == 'C']\n",
    "            new_exon = exon_region[8:] #as first 8 characters are gene description\n",
    "\n",
    "            #counting the number of nucleotide\n",
    "            a_exon= 0\n",
    "            t_exon = 0\n",
    "            g_exon = 0\n",
    "            c_exon = 0\n",
    "\n",
    "            for nt in new_exon:\n",
    "                if nt == 'A':\n",
    "                    a_exon += 1\n",
    "                elif nt == 'T':\n",
    "                    t_exon += 1\n",
    "                elif nt == 'G':\n",
    "                    g_exon += 1\n",
    "                elif nt == 'C':\n",
    "                    c_exon += 1\n",
    "\n",
    "            #summation of all nucleotide lenghts for finding the lenght of exon\n",
    "            total_exon = a_exon + t_exon + g_exon + c_exon\n",
    "            \n",
    "            #C-G ratio of the exon\n",
    "            c_g_exon = ((c_exon + g_exon) / total_exon) * 100\n",
    "\n",
    "            ex = {'count':exon_num, 'A':a_exon, 'C':c_exon, 'T':t_exon, 'G':g_exon, 'Total':total_exon,\n",
    "                 'C-G-ratio':c_g_exon}\n",
    "\n",
    "            exon_list.append(ex)\n",
    "\n",
    "        else:\n",
    "            intron_num += 1  \n",
    "            intr = {}\n",
    "            \n",
    "            intron_region = [char for char in item[1] if char == 'A' or char == 'T' or char == 'G' or char == 'C']\n",
    "            new_intron = intron_region[6:]\n",
    "            \n",
    "            a_intron = 0\n",
    "            t_intron = 0\n",
    "            g_intron = 0\n",
    "            c_intron = 0 \n",
    "            \n",
    "            for nt in new_intron:\n",
    "                if nt == 'A':\n",
    "                    a_intron += 1\n",
    "                elif nt == 'T':\n",
    "                    t_intron += 1\n",
    "                elif nt == 'G':\n",
    "                    g_intron += 1\n",
    "                elif nt == 'C':\n",
    "                    c_intron += 1\n",
    "                    \n",
    "            total_intron = a_intron + t_intron + g_intron + c_intron\n",
    "            \n",
    "            c_g_intron = ((c_intron + g_intron) / total_intron) * 100\n",
    "            \n",
    "            intr = {'count':intron_num, 'A':a_intron, 'C':c_intron, 'T':t_intron, 'G':g_intron, 'Total':total_intron,\n",
    "                   'C-G-ratio':c_g_intron}\n",
    "            \n",
    "            intron_list.append(intr)\n",
    "\n",
    "        \n",
    "    print(exon_list)\n",
    "    print(intron_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Other parameter required for summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#total exon lenght\n",
    "exon_lenght = 0\n",
    "a_exon = 0\n",
    "c_exon = 0\n",
    "t_exon = 0\n",
    "g_exon = 0\n",
    "for i in exon_list:\n",
    "    exon_lenght +=  i['Total']\n",
    "    a_exon += i['A']\n",
    "    c_exon += i['C']\n",
    "    t_exon += i['T']\n",
    "    g_exon += i['G']\n",
    "\n",
    "#total intron lenght\n",
    "intron_lenght = 0\n",
    "a_intron = 0\n",
    "c_intron = 0\n",
    "t_intron = 0\n",
    "g_intron = 0\n",
    "for j in intron_list:\n",
    "    intron_lenght += j['Total']\n",
    "    a_intron += j['A']\n",
    "    c_intron += j['C']\n",
    "    t_intron += j['T']\n",
    "    g_intron += j['G']\n",
    "\n",
    "#total gene lenght\n",
    "gene_lenght = exon_lenght + intron_lenght\n",
    "\n",
    "#sum of all nucleotides individually in the gene\n",
    "a = a_exon + a_intron\n",
    "t = t_exon + t_intron\n",
    "g = g_exon + g_intron\n",
    "c = c_exon + c_intron"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<h3>Writing to a file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('Summary.txt', 'w') as result2:\n",
    "    result2.write(\"\\nCYCS Dog gene description:\")\n",
    "    result2.write(\"\\nThe gene lenght is : \" + str(gene_lenght) + (\" base pairs\"))\n",
    "    result2.write(\"\\n\\nNuleotide Count in gene\\n---------------------------\")\n",
    "    result2.write(\"\\nNo. of A's is :\" + str(a))\n",
    "    result2.write(\"\\nNo.of T's is : \" + str(t))\n",
    "    result2.write(\"\\nNo. of G's is : \" + str(g))\n",
    "    result2.write(\"\\nNo. of C's is : \" + str(c)) \n",
    "    result2.write(\"\\nThe C-G Ratio is : \" + str(((c + g) / (gene_lenght)) * 100) + (\"%\"))\n",
    "    \n",
    "    result2.write(\"\\n\\n No. of exons is : \" + str(exon_num))\n",
    "    result2.write(\"\\nLenght of exon region is : \" + str(exon_lenght) + (\" base pairs\"))\n",
    "    \n",
    "    for i in range(len(exon_list)):\n",
    "        result2.write(\"\\n\\nExon \" + str(i) + \"\\n----------------\")\n",
    "        result2.write(\"\\nNo. of A's is : \" + str(exon_list[i]['A']))\n",
    "        result2.write(\"\\nNo.of T's is : \" + str(exon_list[i]['T']))\n",
    "        result2.write(\"\\nNo. of G's is : \" + str(exon_list[i]['G']))\n",
    "        result2.write(\"\\nNo. of C's is : \" + str(exon_list[i]['C']))\n",
    "        result2.write(\"\\nLenght of exon is : \" + str(exon_list[i]['Total']) + (\" base pairs\"))\n",
    "        result2.write(\"\\nThe C-G Ratio is : \" + str(round(exon_list[i]['C-G-ratio'],2)) + \"%\")\n",
    "                 \n",
    "    result2.write(\"\\n\\n No. of introns is : \" +str(intron_num))\n",
    "    result2.write(\"\\nLenght of intron region is : \" + str(intron_lenght) + (\" base pairs\"))\n",
    "    \n",
    "    for j in range(len(intron_list)):\n",
    "        result2.write(\"\\n\\nIntron \" + str(j) + \"\\n----------------\")\n",
    "        result2.write(\"\\nNo. of A's is : \" + str(intron_list[j]['A']))\n",
    "        result2.write(\"\\nNo.of T's is : \" + str(intron_list[j]['T']))\n",
    "        result2.write(\"\\nNo. of G's is : \" + str(intron_list[j]['G']))\n",
    "        result2.write(\"\\nNo. of C's is : \" + str(intron_list[j]['C']) )\n",
    "        result2.write(\"\\nLenght of intron is : \" + str(intron_list[j]['Total']) + (\" base pairs\"))\n",
    "        result2.write(\"\\nThe C-G Ratio is : \" + str(round(intron_list[j]['C-G-ratio'],2)) + \"%\")\n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
